R/driver_rds_remote.R

```diff
@@ -0,0 +1,199 @@
+      file_ops$create_dir(path = path)
+      file_ops$create_dir(path = file.path(path, "data"))
+      file_ops$create_dir(path = file.path(path, "keys"))
+      file_ops$create_dir(path = file.path(path, "config"))
+
+      self$file_ops <- file_ops
+      self$path <- path
+
+      ## This is a bit of complicated dancing around to mantain
+      ## backward compatibility while allowing better defaults in
+      ## future versions.  I'm writing out a version number here that
+      ## future versions of driver_rds can use to patch, warn or
+      ## change behaviour with older versions of the storr.
+      if (!is_new && !file_ops$object_exists(path = storr:::driver_rds_config_file(path, "version"))) {
+        write_if_missing("1.0.1", path = storr:::driver_rds_config_file(path, "version"), file_ops = file_ops)
+        write_if_missing("TRUE", path =  storr:::driver_rds_config_file(path, "mangle_key_pad"), file_ops = file_ops)
+        write_if_missing("TRUE", path =  storr:::driver_rds_config_file(path, "compress"), file_ops = file_ops)
+        write_if_missing("md5", path =  storr:::driver_rds_config_file(path, "hash_algorithm"), file_ops = file_ops)
+      }
+      ## Then write out the version number:
+      write_if_missing(value = as.character(packageVersion("storr")),
+                       path = storr:::driver_rds_config_file(path, "version"),
+                       file_ops = file_ops)
+
+      if (!is.null(mangle_key)) {
+        storr:::assert_scalar_logical(mangle_key)
+      }
+      self$mangle_key <- driver_rds_config(file_ops, path, "mangle_key", mangle_key,
+                                           FALSE, TRUE)
+
+      if (!is.null(mangle_key_pad)) {
+        storr:::assert_scalar_logical(mangle_key_pad)
+      }
+      self$mangle_key_pad <-driver_rds_config(file_ops, path, "mangle_key_pad",
+                                              mangle_key_pad, FALSE, TRUE)
+
+      if (!is.null(compress)) {
+        storr:::assert_scalar_logical(compress)
+      }
+      self$compress <- driver_rds_config(file_ops, path, "compress", compress,
+                                         TRUE, FALSE)
+
+      if (!is.null(hash_algorithm)) {
+        storr:::assert_scalar_character(hash_algorithm)
+      }
+      self$hash_algorithm <- driver_rds_config(file_ops, path, "hash_algorithm",
+                                               hash_algorithm, "md5", TRUE)
+    },
+
+    type = function() {
+      "s3"
+    },
+
+    destroy = function() {
+      self$file_ops$delete_recursive(path = self$path)
+    },
+
+    get_hash = function(key, namespace) {
+      self$file_ops$readLines(path = self$name_key(key, namespace))
+    },
+
+    set_hash = function(key, namespace, hash) {
+      self$file_ops$create_dir(path = self$name_key("", namespace))
+      self$file_ops$writeLines(text = hash, path = self$name_key(key, namespace))
+      #*** should be making use of (or making an equivalent version of) the
```

Comment:
This is something that should be done in the file_ops class I think

Reply:
My intention in the file ops object was to boil down the fewest set of file operations needed (list directory, read file, write file, delete file, delete directory etc.), so that non-storr related code (specific to a back end) could be encapsulated within that and nowhere else. I was then _using_ those minimal functions to recreate the other driver objects.  My thinking is that other backends/storage types just need to replicate the object with a minimal set of file operations and plug it into the driver_rds_remote object.  I still don't really know the storr codebase well enough to know if that is a good strategy or not... Any thoughts?

Reply:
This is a _fabulous_ strategy - it's absolutely the right way of doing things and something I'd not thought of.  But these can still be made atomic as much as possible

Reply:
Can I just re-iterate how much I like your idea